File: boards/esp32-s3-devkitc1/src/LEDMatrix.cpp
#include "LEDMatrix.h"

Adafruit_NeoPixel ledMatrix(NUM_LEDS, LED_MATRIX_PIN, NEO_GRB + NEO_KHZ800);

void initializeLEDMatrix() {
  ledMatrix.begin();
  ledMatrix.setBrightness(100);
  turnOffLEDMatrix();
}

void setLEDMatrixPixel(uint8_t index, uint8_t red, uint8_t green, uint8_t blue) {
  if (index < NUM_LEDS) {
    ledMatrix.setPixelColor(index, ledMatrix.Color(red, green, blue));
  }
}

void turnOffLEDMatrix() {
  for (uint8_t i = 0; i < NUM_LEDS; i++) {
    ledMatrix.setPixelColor(i, ledMatrix.Color(0, 0, 0));
  }
  ledMatrix.show();
}

void showLEDMatrix() {
  ledMatrix.show();
}

Adafruit_NeoPixel* getLEDMatrix() {
  return &ledMatrix;
}

void runLEDMatrixTest() {
  turnOffLEDMatrix();
  delay(100);
  
  for (uint8_t i = 0; i < NUM_LEDS; i++) {
    uint8_t pos = (i * 256 / NUM_LEDS);
    uint8_t red, green, blue;
    
    if (pos < 85) {
      red = pos * 3;
      green = 255 - pos * 3;
      blue = 0;
    } else if (pos < 170) {
      pos -= 85;
      red = 255 - pos * 3;
      green = 0;
      blue = pos * 3;
    } else {
      pos -= 170;
      red = 0;
      green = pos * 3;
      blue = 255 - pos * 3;
    }
    
    setLEDMatrixPixel(i, red, green, blue);
    showLEDMatrix();
    delay(200);
  }
  
  turnOffLEDMatrix();
}

File: blocklyduino/generators/arduino/playground_master/ledmatrix.js
/**
 * Visual Blocks Language
 *
 * Copyright 2024
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @fileoverview Generating Arduino for LED Matrix blocks.
 */
'use strict';

Blockly.Arduino.led_matrix_init = function() {
  Blockly.Arduino.definitions_['include_ledmatrix_h'] = '#include "LEDMatrix.h"\n';
  var code = 'initializeLEDMatrix();  // Initialize 5x5 LED matrix display\n';
  return code;
};

Blockly.Arduino.led_matrix_set_pixel = function() {
  Blockly.Arduino.definitions_['include_ledmatrix_h'] = '#include "LEDMatrix.h"\n';
  var index = Blockly.Arduino.valueToCode(this, 'INDEX', Blockly.Arduino.ORDER_ATOMIC);
  var red = Blockly.Arduino.valueToCode(this, 'RED', Blockly.Arduino.ORDER_ATOMIC);
  var green = Blockly.Arduino.valueToCode(this, 'GREEN', Blockly.Arduino.ORDER_ATOMIC);
  var blue = Blockly.Arduino.valueToCode(this, 'BLUE', Blockly.Arduino.ORDER_ATOMIC);
  var code = 'setLEDMatrixPixel(' + index + ', ' + red + ', ' + green + ', ' + blue + ');  // Set pixel ' + index + ' color\n';
  return code;
};

Blockly.Arduino.led_matrix_fill = function() {
  Blockly.Arduino.definitions_['include_ledmatrix_h'] = '#include "LEDMatrix.h"\n';
  var red = Blockly.Arduino.valueToCode(this, 'RED', Blockly.Arduino.ORDER_ATOMIC);
  var green = Blockly.Arduino.valueToCode(this, 'GREEN', Blockly.Arduino.ORDER_ATOMIC);
  var blue = Blockly.Arduino.valueToCode(this, 'BLUE', Blockly.Arduino.ORDER_ATOMIC);
  var code = 'fillLEDMatrix(' + red + ', ' + green + ', ' + blue + ');  // Fill entire matrix with color\n';
  return code;
};

Blockly.Arduino.led_matrix_show = function() {
  Blockly.Arduino.definitions_['include_ledmatrix_h'] = '#include "LEDMatrix.h"\n';
  var code = 'showLEDMatrix();  // Display the current buffer on the matrix\n';
  return code;
};

Blockly.Arduino.led_matrix_off = function() {
  Blockly.Arduino.definitions_['include_ledmatrix_h'] = '#include "LEDMatrix.h"\n';
  var code = 'turnOffLEDMatrix();  // Turn off all LEDs in the matrix\n';
  return code;
};

Blockly.Arduino.led_matrix_test = function() {
  Blockly.Arduino.definitions_['include_ledmatrix_h'] = '#include "LEDMatrix.h"\n';
  var code = 'runLEDMatrixTest();\n';
  return code;
};
